src/release_entry.rs

```diff
@@ -19,21 +26,103 @@ pub struct ReleaseEntry {
   pub sha256: [u8; 32],
   pub filename_or_url: String,
   pub version: Version,
-  pub length: i64,
+  pub length: u64,
   pub is_delta: bool,
   pub percentage: i32,
 }
 
-impl Default for ReleaseEntry {
-  fn default() -> ReleaseEntry {
-    ReleaseEntry {
-      filename_or_url: "Foobdar".to_owned(),
-      version: Version::parse("1.0.0").unwrap(),
-      is_delta: true,
-      length: 42,
+static HEADER: &'static str = "# SHA256 of the file                                             Name       Version Size  [delta/full] release%";
+
+//
+// Create new release entries
+impl ReleaseEntry {
+  pub fn from_file(path: &str) -> Result<ReleaseEntry, Box<Error>> {
+    let mut sha256_sum = Sha256::new();
+    {
+      let file = try!(File::open(path));
+      let mut content: [u8; 65536] = [0; 65536];
+      let mut buf = BufReader::new(file);
+
+      loop {
+        let count = try!(buf.read(&mut content));
+        if count == 0 { break; }
+
+        sha256_sum.input(&content[..count]);
+      }
+    }
+
+    let p = Path::new(path);
+    let stat = try!(fs::metadata(path));
+    let file = p.file_name().unwrap();
+
+    let mut ret = ReleaseEntry {
       sha256: [0; 32],
+      filename_or_url: file.to_str().unwrap().to_owned(),
+      length: stat.len(),
+      version: Version::parse("0.0.0").unwrap(),
+      is_delta: false,
       percentage: 100,
+    };
+
+    let sha = sha256_sum.result();
+    for i in 0..32 {
+      ret.sha256[i] = sha[i];
     }
+
+    return Ok(ret);
+  }
+}
+
+//
+// Write release files
+//
+
+impl ReleaseEntry {
+  fn sha256_to_string(&self) -> String {
+    return self.sha256.into_iter().fold(
+      "".to_owned(), 
+      |mut acc, x| { 
+        let byte = format!("{:02x}", x);
+        acc.push_str(&byte); 
+        return acc; 
+      });
+  }
+
+  fn escape_filename(&self) -> String {
```

Comment:
Not sure how much you care about allocations here, but you might find the [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html) type interesting. It would let you express a return type here that might be borrowed (in case `SCHEME.is_match`) or owned (in case it needs to be encoded).

It would basically look like:

```rust
fn escape_filename(&self) -> Cow<str> {
    if match {
        Cow::Borrowed(&self.filename_or_url)
    } else {
        Cow::Owned(utf8_percent_encode(&self.filename_or_url))
    }
}
```